core/common/src/YearMonth.kt

```diff
@@ -0,0 +1,396 @@
+ *
+ * ### Construction, serialization, and deserialization
+ *
+ * [YearMonth] can be constructed directly from its components using the constructor.
+ * See sample 1.
+ *
+ * [parse] and [toString] methods can be used to obtain a [YearMonth] from and convert it to a string in the
+ * ISO 8601 extended format.
+ * See sample 2.
+ *
+ * [parse] and [YearMonth.format] both support custom formats created with [Format] or defined in [Formats].
+ * See sample 3.
+ *
+ * Additionally, there are several `kotlinx-serialization` serializers for [YearMonth]:
+ * - [YearMonthIso8601Serializer] for the ISO 8601 extended format.
+ * - [YearMonthComponentSerializer] for an object with components.
+ *
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.constructorFunctionMonthNumber
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.simpleParsingAndFormatting
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.customFormat
+ */
+@Serializable(with = YearMonthIso8601Serializer::class)
+public class YearMonth
+/**
+ * Constructs a [YearMonth] instance from the given year-month components.
+ *
+ * The [month] component is 1-based.
+ *
+ * The supported ranges of components:
+ * - [year] the range is platform-dependent, but at least is enough to represent year-months of all instants between
+ *          [Instant.DISTANT_PAST] and [Instant.DISTANT_FUTURE] in any time zone.
+ * - [month] `1..12`
+ *
+ * @throws IllegalArgumentException if any parameter is out of range.
+ * @sample kotlinx.datetime.test.samples.YearMonthSamples.constructorFunctionMonthNumber
+ */
+public constructor(year: Int, month: Int) : Comparable<YearMonth> {
+    /**
+     * Returns the year component of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.year
+     */
+    public val year: Int = year
+
+    /**
+     * Returns the number-of-the-month (1..12) component of the year-month.
+     *
+     * Shortcut for `month.number`.
+     */
+    internal val monthNumber: Int = month
+
+    init {
+        require(month in 1..12) { "Month must be in 1..12, but was $month" }
+        require(year in LocalDate.MIN.year..LocalDate.MAX.year) {
+            "Year $year is out of range: ${LocalDate.MIN.year}..${LocalDate.MAX.year}"
+        }
+    }
+
+    /**
+     * Returns the month ([Month]) component of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.month
+     */
+    public val month: Month get() = Month(monthNumber)
+
+    /**
+     * Returns the first day of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.firstAndLastDay
+     */
+    public val firstDay: LocalDate get() = onDay(1)
+
+    /**
+     * Returns the last day of the year-month.
+     *
+     * @sample kotlinx.datetime.test.samples.YearMonthSamples.firstAndLastDay
+     */
+    public val lastDay: LocalDate get() = onDay(numberOfDays)
```

Comment:
You should add contains methods for LocalDate and LocalDateTime as in:
```
    public fun contains(date: LocalDate): Boolean = date in firstDay..lastDay
```

Also helpful would be intersects functions for ranges to determine if this year month has any overlap with another range. This should include the cases of `ClosedRange<LocalDate>` and `OpenEndRange<LocalDateTime>`. 

```
    public fun intersects(dateRange: ClosedRange<LocalDate>): Boolean = ...
    public fun intersects(dateTimeRange: OpenEndRange<LocalDateTime>): Boolean = ...
```